// whaileys/lib/Utils/process-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldIncrementChatUnread = exports.isRealMessage = exports.decryptSecretEncryptedMessage = exports.cleanMessage = void 0;
const WAProto_1 = require("../../WAProto");
const Types_1 = require("../Types");
const Utils_1 = require("../Utils");
const WABinary_1 = require("../WABinary");
const reporting_utils_1 = require("./reporting-utils");
const SECRET_ENC_LABELS = {
    [WAProto_1.proto.Message.SecretEncryptedMessage.SecretEncType.MESSAGE_EDIT]: "Message Edit",
    [WAProto_1.proto.Message.SecretEncryptedMessage.SecretEncType.EVENT_EDIT]: "Event Edit"
};
const MSG_MISSED_CALL_TYPES = new Set([
    Types_1.WAMessageStubType.CALL_MISSED_GROUP_VIDEO,
    Types_1.WAMessageStubType.CALL_MISSED_GROUP_VOICE,
    Types_1.WAMessageStubType.CALL_MISSED_VIDEO,
    Types_1.WAMessageStubType.CALL_MISSED_VOICE
]);
/** Cleans a received message to further processing */
const cleanMessage = (message, meId, meLid) => {
    var _a, _b;
    // ensure remoteJid and participant doesn't have device or agent in it
    message.key.remoteJid = (0, WABinary_1.jidNormalizedUser)(message.key.remoteJid);
    message.key.participant = message.key.participant
        ? (0, WABinary_1.jidNormalizedUser)(message.key.participant)
        : undefined;
    const content = (0, Utils_1.normalizeMessageContent)(message.message);
    // if the message has a reaction, ensure fromMe & remoteJid are from our perspective
    if ((_a = content === null || content === void 0 ? void 0 : content.reactionMessage) === null || _a === void 0 ? void 0 : _a.key) {
        normaliseKey(content.reactionMessage.key);
    }
    if ((_b = content === null || content === void 0 ? void 0 : content.secretEncryptedMessage) === null || _b === void 0 ? void 0 : _b.targetMessageKey) {
        normaliseKey(content.secretEncryptedMessage.targetMessageKey);
    }
    function normaliseKey(msgKey) {
        // if the reaction is from another user
        // we've to correctly map the key to this user's perspective
        if (!message.key.fromMe) {
            // if the sender believed the message being reacted to is not from them
            // we've to correct the key to be from them, or some other participant
            msgKey.fromMe = !msgKey.fromMe
                ? (0, WABinary_1.areJidsSameUser)(msgKey.participant || msgKey.remoteJid, meId) ||
                    (!!meLid &&
                        (0, WABinary_1.areJidsSameUser)(msgKey.participant || msgKey.remoteJid, meLid))
                : // if the message being reacted to, was from them
                    // fromMe automatically becomes false
                    false;
            // set the remoteJid to being the same as the chat the message came from
            msgKey.remoteJid = message.key.remoteJid;
            // set participant of the message
            msgKey.participant = msgKey.participant || message.key.participant;
        }
    }
};
exports.cleanMessage = cleanMessage;
const decryptSecretEncryptedMessage = async (message, messageSecret, meId, meLid, logger) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const content = (0, Utils_1.normalizeMessageContent)(message.message);
    const secretEncryptedMessage = content === null || content === void 0 ? void 0 : content.secretEncryptedMessage;
    if (!secretEncryptedMessage)
        return;
    const targetMessageKey = secretEncryptedMessage.targetMessageKey;
    const secretEncType = secretEncryptedMessage.secretEncType;
    const { SecretEncType } = WAProto_1.proto.Message.SecretEncryptedMessage;
    if (secretEncType === null ||
        secretEncType === undefined ||
        secretEncType === SecretEncType.UNKNOWN ||
        !(targetMessageKey === null || targetMessageKey === void 0 ? void 0 : targetMessageKey.id)) {
        logger === null || logger === void 0 ? void 0 : logger.warn({
            secretEncType,
            targetMessageKey: secretEncryptedMessage.targetMessageKey
        }, "unsupported secret encrypted message type");
        return;
    }
    const useCaseLabel = SECRET_ENC_LABELS[secretEncType];
    if (!useCaseLabel) {
        logger === null || logger === void 0 ? void 0 : logger.info({ secretEncType, targetMessageKey, messageKey: message.key }, "no HKDF label registered for secret encrypted message type, leaving envelope intact");
        return;
    }
    if (!((_a = secretEncryptedMessage.encPayload) === null || _a === void 0 ? void 0 : _a.length) ||
        !((_b = secretEncryptedMessage.encIv) === null || _b === void 0 ? void 0 : _b.length)) {
        logger === null || logger === void 0 ? void 0 : logger.warn({ targetMessageKey }, "missing encrypted edit payload");
        return;
    }
    const modLid = message.key.senderLid || message.key.participantLid;
    const targetLid = targetMessageKey.senderLid || targetMessageKey.participantLid;
    const envelopeIsLid = !!modLid ||
        !!targetLid ||
        (0, WABinary_1.isLidUser)((_c = message.key.remoteJid) !== null && _c !== void 0 ? _c : undefined) ||
        (0, WABinary_1.isLidUser)((_d = message.key.participant) !== null && _d !== void 0 ? _d : undefined) ||
        (0, WABinary_1.isLidUser)((_e = targetMessageKey.remoteJid) !== null && _e !== void 0 ? _e : undefined) ||
        (0, WABinary_1.isLidUser)((_f = targetMessageKey.participant) !== null && _f !== void 0 ? _f : undefined);
    const ownSender = (0, WABinary_1.jidNormalizedUser)(envelopeIsLid && meLid ? meLid : meId);
    const originalSender = targetMessageKey.fromMe
        ? ownSender
        : (0, WABinary_1.jidNormalizedUser)((envelopeIsLid && (targetLid || modLid)) ||
            targetMessageKey.participant ||
            targetMessageKey.remoteJid ||
            "");
    const modificationSender = message.key.fromMe
        ? ownSender
        : (0, WABinary_1.jidNormalizedUser)((envelopeIsLid && modLid) ||
            message.key.participant ||
            message.key.remoteJid ||
            "");
    if (!originalSender || !modificationSender) {
        logger === null || logger === void 0 ? void 0 : logger.warn({ targetMessageKey, messageKey: message.key }, "missing sender for secret encrypted message");
        return;
    }
    let editedMessage;
    try {
        const decryptKey = await (0, reporting_utils_1.generateMsgSecretKey)(useCaseLabel, targetMessageKey.id, originalSender, modificationSender, messageSecret);
        const decrypted = (0, Utils_1.aesDecryptGCM)(secretEncryptedMessage.encPayload, decryptKey, secretEncryptedMessage.encIv, Buffer.alloc(0));
        editedMessage = WAProto_1.proto.Message.decode(decrypted);
    }
    catch (err) {
        logger === null || logger === void 0 ? void 0 : logger.warn({
            err,
            targetMessageKey,
            messageKey: message.key,
            originalSender,
            modificationSender
        }, "failed to decrypt secret encrypted message");
        return;
    }
    if (((_g = editedMessage.protocolMessage) === null || _g === void 0 ? void 0 : _g.type) ===
        WAProto_1.proto.Message.ProtocolMessage.Type.MESSAGE_EDIT) {
        message.message = editedMessage;
        return;
    }
    if (((_h = message.message) === null || _h === void 0 ? void 0 : _h.messageContextInfo) &&
        !editedMessage.messageContextInfo) {
        editedMessage.messageContextInfo = message.message.messageContextInfo;
    }
    message.message = {
        protocolMessage: {
            key: targetMessageKey,
            type: WAProto_1.proto.Message.ProtocolMessage.Type.MESSAGE_EDIT,
            editedMessage,
            timestampMs: (0, Utils_1.toNumber)(message.messageTimestamp) * 1000
        }
    };
};
exports.decryptSecretEncryptedMessage = decryptSecretEncryptedMessage;
const isRealMessage = (message) => {
    const normalizedContent = (0, Utils_1.normalizeMessageContent)(message.message);
    return ((!!normalizedContent ||
        MSG_MISSED_CALL_TYPES.has(message.messageStubType)) &&
        !(normalizedContent === null || normalizedContent === void 0 ? void 0 : normalizedContent.protocolMessage) &&
        !(normalizedContent === null || normalizedContent === void 0 ? void 0 : normalizedContent.secretEncryptedMessage) &&
        !(normalizedContent === null || normalizedContent === void 0 ? void 0 : normalizedContent.reactionMessage));
};
exports.isRealMessage = isRealMessage;
const shouldIncrementChatUnread = (message) => !message.key.fromMe && !message.messageStubType;
exports.shouldIncrementChatUnread = shouldIncrementChatUnread;
const updateGroupMetadata = (groupMetadata, action, participants) => {
    const findParticipant = (id) => groupMetadata.participants.find(p => (0, WABinary_1.areJidsSameUser)(p.id, id));
    switch (action) {
        case "add":
            participants.forEach(id => {
                if (!findParticipant(id)) {
                    groupMetadata.participants.push({ id, admin: null });
                }
            });
            break;
        case "remove":
            groupMetadata.participants = groupMetadata.participants.filter(p => !participants.includes(p.id));
            break;
        case "promote":
            participants.forEach(id => {
                const p = findParticipant(id);
                if (p)
                    p.admin = "admin";
            });
            break;
        case "demote":
            participants.forEach(id => {
                const p = findParticipant(id);
                if (p)
                    p.admin = null;
            });
            break;
    }
    return groupMetadata;
};
const processMessage = async (message, { shouldProcessHistoryMsg, ev, creds, keyStore, logger, options, config }) => {
    var _a, _b, _c, _d, _e, _f;
    const meId = creds.me.id;
    const { accountSettings } = creds;
    const chat = { id: (0, WABinary_1.jidNormalizedUser)(message.key.remoteJid) };
    if ((0, exports.isRealMessage)(message)) {
        chat.conversationTimestamp = (0, Utils_1.toNumber)(message.messageTimestamp);
        // only increment unread count if not CIPHERTEXT and from another person
        if ((0, exports.shouldIncrementChatUnread)(message)) {
            chat.unreadCount = (chat.unreadCount || 0) + 1;
        }
        if (accountSettings === null || accountSettings === void 0 ? void 0 : accountSettings.unarchiveChats) {
            chat.archived = false;
            chat.readOnly = false;
        }
    }
    const content = (0, Utils_1.normalizeMessageContent)(message.message);
    const protocolMsg = content === null || content === void 0 ? void 0 : content.protocolMessage;
    if (protocolMsg) {
        switch (protocolMsg.type) {
            case WAProto_1.proto.Message.ProtocolMessage.Type.HISTORY_SYNC_NOTIFICATION:
                const histNotification = protocolMsg.historySyncNotification;
                const process = shouldProcessHistoryMsg;
                const isLatest = !((_a = creds.processedHistoryMessages) === null || _a === void 0 ? void 0 : _a.length);
                logger === null || logger === void 0 ? void 0 : logger.info({
                    histNotification,
                    process,
                    id: message.key.id,
                    isLatest
                }, "got history notification");
                if (process) {
                    if (histNotification.syncType !==
                        WAProto_1.proto.HistorySync.HistorySyncType.ON_DEMAND) {
                        ev.emit("creds.update", {
                            processedHistoryMessages: [
                                ...(creds.processedHistoryMessages || []),
                                { key: message.key, messageTimestamp: message.messageTimestamp }
                            ]
                        });
                    }
                    const data = await (0, Utils_1.downloadAndProcessHistorySyncNotification)(histNotification, options);
                    ev.emit("messaging-history.set", {
                        ...data,
                        isLatest: histNotification.syncType !==
                            WAProto_1.proto.HistorySync.HistorySyncType.ON_DEMAND
                            ? isLatest
                            : false
                    });
                }
                break;
            case WAProto_1.proto.Message.ProtocolMessage.Type.APP_STATE_SYNC_KEY_SHARE:
                const keys = protocolMsg.appStateSyncKeyShare.keys;
                if (keys === null || keys === void 0 ? void 0 : keys.length) {
                    let newAppStateSyncKeyId = "";
                    await keyStore.transaction(async () => {
                        const newKeys = [];
                        for (const { keyData, keyId } of keys) {
                            const strKeyId = Buffer.from(keyId.keyId).toString("base64");
                            newKeys.push(strKeyId);
                            await keyStore.set({
                                "app-state-sync-key": { [strKeyId]: keyData }
                            });
                            newAppStateSyncKeyId = strKeyId;
                        }
                        logger === null || logger === void 0 ? void 0 : logger.info({ newAppStateSyncKeyId, newKeys }, "injecting new app state sync keys");
                    });
                    ev.emit("creds.update", { myAppStateKeyId: newAppStateSyncKeyId });
                }
                else {
                    logger === null || logger === void 0 ? void 0 : logger.info({ protocolMsg }, "recv app state sync with 0 keys");
                }
                break;
            case WAProto_1.proto.Message.ProtocolMessage.Type.REVOKE:
                ev.emit("messages.update", [
                    {
                        key: {
                            ...message.key,
                            id: protocolMsg.key.id
                        },
                        update: {
                            message: null,
                            messageStubType: Types_1.WAMessageStubType.REVOKE,
                            key: message.key
                        }
                    }
                ]);
                break;
            case WAProto_1.proto.Message.ProtocolMessage.Type.EPHEMERAL_SETTING:
                Object.assign(chat, {
                    ephemeralSettingTimestamp: (0, Utils_1.toNumber)(message.messageTimestamp),
                    ephemeralExpiration: protocolMsg.ephemeralExpiration || null
                });
                if (chat.id && (0, WABinary_1.isJidGroup)(chat.id)) {
                    const groupMetadata = (_b = config.groupMetadataCache) === null || _b === void 0 ? void 0 : _b.get(chat.id);
                    if (!groupMetadata)
                        break;
                    (_c = config.groupMetadataCache) === null || _c === void 0 ? void 0 : _c.set(chat.id, {
                        ...groupMetadata,
                        ephemeralDuration: protocolMsg.ephemeralExpiration || null
                    });
                }
                break;
            case WAProto_1.proto.Message.ProtocolMessage.Type
                .PEER_DATA_OPERATION_REQUEST_RESPONSE_MESSAGE:
                const response = protocolMsg.peerDataOperationRequestResponseMessage;
                if (!(response === null || response === void 0 ? void 0 : response.peerDataOperationResult))
                    break;
                const { peerDataOperationResult } = response;
                const responseMessages = peerDataOperationResult
                    .map(result => {
                    var _a;
                    const responseBytes = (_a = result.placeholderMessageResendResponse) === null || _a === void 0 ? void 0 : _a.webMessageInfoBytes;
                    if (!responseBytes)
                        return;
                    return WAProto_1.proto.WebMessageInfo.decode(responseBytes);
                })
                    .filter(Boolean);
                ev.emit("messages.pdo-response", {
                    messages: responseMessages
                });
                break;
            case WAProto_1.proto.Message.ProtocolMessage.Type.SHARE_PHONE_NUMBER:
                ev.emit("chats.phoneNumberShare", {
                    lid: message.key.senderLid,
                    jid: message.key.senderPn
                });
                ev.emit("contacts.phone-number-share", {
                    lid: message.key.senderLid,
                    jid: message.key.senderPn
                });
                break;
        }
    }
    else if (content === null || content === void 0 ? void 0 : content.reactionMessage) {
        const reaction = {
            ...content.reactionMessage,
            key: message.key
        };
        ev.emit("messages.reaction", [
            {
                reaction,
                key: content.reactionMessage.key
            }
        ]);
    }
    else if (message.messageStubType) {
        const jid = message.key.remoteJid;
        //let actor = whatsappID (message.participant)
        let participants;
        const emitParticipantsUpdate = (action) => {
            var _a, _b;
            ev.emit("group-participants.update", { id: jid, participants, action });
            const groupMetadata = (_a = config.groupMetadataCache) === null || _a === void 0 ? void 0 : _a.get(jid);
            if (!groupMetadata)
                return;
            const updatedMetadata = updateGroupMetadata(groupMetadata, action, participants);
            (_b = config.groupMetadataCache) === null || _b === void 0 ? void 0 : _b.set(jid, updatedMetadata);
        };
        const emitGroupUpdate = (update) => {
            var _a, _b;
            ev.emit("groups.update", [{ id: jid, ...update }]);
            const groupMetadata = (_a = config.groupMetadataCache) === null || _a === void 0 ? void 0 : _a.get(jid);
            if (!groupMetadata)
                return;
            (_b = config.groupMetadataCache) === null || _b === void 0 ? void 0 : _b.set(jid, { ...groupMetadata, ...update });
        };
        const participantsIncludesMe = () => participants.find(jid => (0, WABinary_1.areJidsSameUser)(meId, jid));
        switch (message.messageStubType) {
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_LEAVE:
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_REMOVE:
                participants = message.messageStubParameters || [];
                emitParticipantsUpdate("remove");
                // mark the chat read only if you left the group
                if (participantsIncludesMe()) {
                    chat.readOnly = true;
                }
                break;
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_ADD:
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_INVITE:
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_ADD_REQUEST_JOIN:
                participants = message.messageStubParameters || [];
                if (participantsIncludesMe()) {
                    chat.readOnly = false;
                }
                emitParticipantsUpdate("add");
                break;
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_DEMOTE:
                participants = message.messageStubParameters || [];
                emitParticipantsUpdate("demote");
                break;
            case Types_1.WAMessageStubType.GROUP_PARTICIPANT_PROMOTE:
                participants = message.messageStubParameters || [];
                emitParticipantsUpdate("promote");
                break;
            case Types_1.WAMessageStubType.GROUP_CHANGE_ANNOUNCE:
                const announceValue = (_d = message.messageStubParameters) === null || _d === void 0 ? void 0 : _d[0];
                emitGroupUpdate({
                    announce: announceValue === "true" || announceValue === "on"
                });
                break;
            case Types_1.WAMessageStubType.GROUP_CHANGE_RESTRICT:
                const restrictValue = (_e = message.messageStubParameters) === null || _e === void 0 ? void 0 : _e[0];
                emitGroupUpdate({
                    restrict: restrictValue === "true" || restrictValue === "on"
                });
                break;
            case Types_1.WAMessageStubType.GROUP_CHANGE_SUBJECT:
                const name = (_f = message.messageStubParameters) === null || _f === void 0 ? void 0 : _f[0];
                chat.name = name;
                emitGroupUpdate({ subject: name });
                break;
        }
    }
    if (Object.keys(chat).length > 1) {
        ev.emit("chats.update", [chat]);
    }
};
exports.default = processMessage;
